ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java

```diff
@@ -157,6 +182,277 @@ public void testMultipleScalingUnit()
+   * @param invalidResponse the AtomicBoolean that will contain whether an unexpected failure was notified.
+   * @param indexToFail the index representing which response for which failure is to be simulated. For example,
+   *                    if index is 0, then the first response will be failed. If the index is -1, no responses will be
+   *                    failed.
+   */
+  private void testDeleteResponseNotification(DeleteManager deleteManager, NetworkClient networkClient,
+      List<ReplicaId> failedReplicaIds, String blobId, AtomicInteger successfulResponseCount,
+      AtomicBoolean invalidResponse, int indexToFail)
+      throws Exception {
+    FutureResult<Void> futureResult = new FutureResult<>();
+    deleteManager.submitDeleteBlobOperation(blobId, futureResult, null);
+    List<RequestInfo> allRequests = new ArrayList<>();
+    while (allRequests.size() < DELETE_REQUEST_PARALLELISM) {
+      deleteManager.poll(allRequests);
+    }
+    ReplicaId replicaIdToFail =
+        indexToFail == -1 ? null : ((RouterRequestInfo) allRequests.get(indexToFail)).getReplicaId();
+    for (RequestInfo requestInfo : allRequests) {
+      if (replicaIdToFail != null && replicaIdToFail.equals(((RouterRequestInfo) requestInfo).getReplicaId())) {
+        mockSelectorState.set(MockSelectorState.DisconnectOnSend);
+      } else {
+        mockSelectorState.set(MockSelectorState.Good);
+      }
+      List<RequestInfo> requestInfoListToSend = new ArrayList<>();
+      requestInfoListToSend.add(requestInfo);
+      List<ResponseInfo> responseInfoList;
+      do {
+        responseInfoList = networkClient.sendAndPoll(requestInfoListToSend, 10);
+        requestInfoListToSend.clear();
+      } while (responseInfoList.size() == 0);
+      deleteManager.handleResponse(responseInfoList.get(0));
+    }
+    futureResult.await();
+    if (indexToFail == -1) {
+      Assert.assertEquals("Successful notification should have arrived for replicas that were up",
+          DELETE_REQUEST_PARALLELISM, successfulResponseCount.get());
+      Assert.assertEquals("Failure detector should not have been notified", 0, failedReplicaIds.size());
+      Assert.assertFalse("There should be no notifications of any other kind", invalidResponse.get());
+    } else {
+      Assert.assertEquals("Failure detector should have been notified", 1, failedReplicaIds.size());
+      Assert.assertEquals("Failed notification should have arrived for the failed replica", replicaIdToFail,
+          failedReplicaIds.get(0));
+      Assert.assertEquals("Successful notification should have arrived for replicas that were up",
+          DELETE_REQUEST_PARALLELISM - 1, successfulResponseCount.get());
+      Assert.assertFalse("There should be no notifications of any other kind", invalidResponse.get());
+    }
+    failedReplicaIds.clear();
+    successfulResponseCount.set(0);
+    invalidResponse.set(false);
+    mockSelectorState.set(MockSelectorState.Good);
+  }
```

Comment:
there is quite a bit of repeated code in these three functions - would recommend trying to make it leaner and reuse code.